### PR TITLE
Release tool improvements

### PR DESCRIPTION
**Pull request by mattclay**
##### SUMMARY

* Add a safety check for missing PyPI artifacts.
* The wheel builds are now repeatable.
* The sdist builds are now repeatable.

##### ISSUE TYPE

Feature Pull Request

##### COMPONENT NAME

packaging/release.py

### CODE REVIEW

**Review by sivel — APPROVED** [review]

**Comment by gotmax23** [comment]
re the repeatable_sdist part:

Ideally, setuptools would natively support reproducible builds and $SOURCE_DATE_EPOCH like hatchling and others (see https://github.com/pypa/setuptools/issues/2133). In the meantime, would it make more sense to move this code to the in tree pyproject build backend?  This function could run after our `build_sdist()` hook calls out to the original `setuptools.build_meta.build_sdist()`.

From the discussions I had with @webknjaz, I understand that the point of removing the Makefile in favor of the in tree build backend was to avoid extra manual steps like this when building the sdist.

Also, it would be nice to have the script notate the SOURCE_DATE_EPOCH alongside the dist checksums in the Github Release template.

**Comment by mattclay** [comment]
@gotmax23 I moved the reproducible tarball processing into the build backend. As for the the `SOURCE_DATE_EPOCH`, that can be retrieved from the release commit. That's really an implementation detail of our release process though. If you're trying to match our sdist exactly using a checkout, running the release tool will accomplish that.

**Review by sivel — APPROVED** [review]

**Review by webknjaz — COMMENTED** [review]
Could you hold this until I merge my PR? Looks like it may have a lot of conflicts with mine.

**Comment by mattclay** [comment]
@webknjaz We were hoping to get this merged before 2.15.0rc2 goes out. It looks like the conflicts should be minimal -- the imports and a minor context issue at the end of build_sdist, both of which should resolve easily.

**Comment by webknjaz** [comment]
@mattclay since I have a series of commits, I'll be faced with multiple instances of that tho. I thought my thing was more or less ready so I was also hoping to get it merged soon.

File: packaging/pep517_backend/_backend.py
Comment (webknjaz) [inline packaging/pep517_backend/_backend.py]: This feels weird — why is the `sdist_directory` disregarded? I'll probably need to check from the laptop. So far, looking from the GH app gives an impression that things might not be right...

File: packaging/pep517_backend/_backend.py
Comment (webknjaz) [inline packaging/pep517_backend/_backend.py]: Function names should use verbs, from this name I can't guess what the side effects of the function are, for example.
Is it "make the incoming sdist reproducible"? "repack the sdist"? Something else?

File: packaging/pep517_backend/_backend.py
Comment (webknjaz) [inline packaging/pep517_backend/_backend.py]: Cap the title sentence to 72 chars per PEP 257.

File: packaging/pep517_backend/_backend.py
Comment (webknjaz) [inline packaging/pep517_backend/_backend.py]: This looks private, meaning it needs an underscore.

Comment (webknjaz) [reply packaging/pep517_backend/_backend.py]: Also a verb.

**Review by webknjaz — COMMENTED** [review]
Looking at this more, I realized I don't understand why this needs to be inside our in-tree backend specifically. This explanation seems to be a missing bit in the PR description. I think I'd like to understand the whys and the unobvious inner workings.

Besides that, I'd expect tests, especially building under the oldest supported Python version and rebuilding under the newest one. I remember the tarfile module had some quirks that we don't even handle in galaxy, for example.

Another thing that concerns me is that this reaches too deep into the gray area of PEP 517 which kinda says not to make other backends weird dependencies and only the setuptools' docs soften the stance by suggesting a thin wrapper. Though, this doesn't seem thin anymore.

**Comment by mattclay** [comment]
@webknjaz Thanks for the suggestion on the function names. I've updated them in the release tool.

**Pull request merged**
